pkg/controller/cachemanager/cachemanager.go

```diff
@@ -0,0 +1,336 @@
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/syncutil"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/syncutil/aggregator"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/target"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/watch"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var log = logf.Log.WithName("cache-manager")
+
+type CacheManagerConfig struct {
+	Opa              syncutil.OpaDataClient
+	SyncMetricsCache *syncutil.MetricsCache
+	Tracker          *readiness.Tracker
+	ProcessExcluder  *process.Excluder
+	Registrar        *watch.Registrar
+	WatchedSet       *watch.Set
+	GVKAggregator    *aggregator.GVKAgreggator
+	Reader           client.Reader
+}
+
+type CacheManager struct {
+	opa              syncutil.OpaDataClient
+	syncMetricsCache *syncutil.MetricsCache
+	tracker          *readiness.Tracker
+	processExcluder  *process.Excluder
+	registrar        *watch.Registrar
+	watchedSet       *watch.Set
+	gvkAggregator    *aggregator.GVKAgreggator
+	gvksToRemove     *watch.Set
+	gvksToSync       *watch.Set
+	replayErrChan    chan error
+	replayTicker     time.Ticker
+	reader           client.Reader
+}
+
+func NewCacheManager(config *CacheManagerConfig) (*CacheManager, error) {
+	if config.WatchedSet == nil {
+		return nil, fmt.Errorf("watchedSet must be non-nil")
+	}
+	if config.Registrar == nil {
+		return nil, fmt.Errorf("registrar must be non-nil")
+	}
+	if config.ProcessExcluder == nil {
+		return nil, fmt.Errorf("processExcluder must be non-nil")
+	}
+	if config.Tracker == nil {
+		return nil, fmt.Errorf("tracker must be non-nil")
+	}
+	if config.Reader == nil {
+		return nil, fmt.Errorf("reader must be non-nil")
+	}
+
+	cm := &CacheManager{
+		opa:              config.Opa,
+		syncMetricsCache: config.SyncMetricsCache,
+		tracker:          config.Tracker,
+		processExcluder:  config.ProcessExcluder,
+		registrar:        config.Registrar,
+		watchedSet:       config.WatchedSet,
+		reader:           config.Reader,
+	}
+
+	cm.gvkAggregator = aggregator.NewGVKAggregator()
+	cm.gvksToRemove = watch.NewSet()
+	cm.gvksToSync = watch.NewSet()
+
+	cm.replayTicker = *time.NewTicker(3 * time.Second)
+
+	return cm, nil
+}
+
+func (c *CacheManager) Start(ctx context.Context) error {
+	go c.updateDatastore(ctx)
+
+	<-ctx.Done()
+	return nil
+}
+
+// WatchGVKsToSync adjusts the watched set of gvks according to the newGVKs passed in
+// for a given {syncSourceType, syncSourceName}.
+func (c *CacheManager) WatchGVKsToSync(ctx context.Context, newGVKs []schema.GroupVersionKind, newExcluder *process.Excluder, syncSourceType, syncSourceName string) error {
+	netNewGVKs := []schema.GroupVersionKind{}
+	for _, gvk := range newGVKs {
+		if !c.gvkAggregator.IsPresent(gvk) {
+			netNewGVKs = append(netNewGVKs, gvk)
+		}
+	}
+	// mark these gvks for the background goroutine to sync
```

Comment:
This couples the background process to this function, which creates more complexity.  In such a model, any future methods like this one would also have to edit gkvsToSync.  We'd rather have these things be decoupled.

A simpler model would be to just be responsible for updating the SOT (gvkAggregator) and then leave it to the background process to:

1. Decide when to run itself
2. Read from the gvkAggregator
3. Decide what actions to take